crates/solidity/outputs/npm/package/src/ast/generated/frontend.ts

```diff
@@ -0,0 +1,4329 @@
+// This file is generated automatically by infrastructure scripts. Please don't edit by hand.
+
+import { ast_internal } from "../../generated";
+import { RuleNode, TokenNode } from "../../cst";
+import { RuleKind, TokenKind } from "../../kinds";
+
+function once<T>(factory: () => T): () => T {
+  let value: T | null = null;
+  return () => {
+    if (value === null) {
+      value = factory();
+    }
+    return value;
+  };
+}
+
+export class SourceUnit {
+  private readonly fields = once(() => ast_internal.pickSequence(this.cst, RuleKind.SourceUnit));
+
+  public constructor(public readonly cst: RuleNode) {}
+
+  public readonly members: () => SourceUnitMembers | null = once(() => {
+    const field = this.fields()[0];
+    return field ? new SourceUnitMembers(field as RuleNode) : null;
+  });
+}
+
+export class PragmaDirective {
+  private readonly fields = once(() => ast_internal.pickSequence(this.cst, RuleKind.PragmaDirective));
+
+  public constructor(public readonly cst: RuleNode) {}
+
+  public readonly pragmaKeyword: () => TokenNode = once(() => {
```

Comment:
I am surprised to see tokens for non-significant keywords and punctuation in the AST. What is the reasoning for including them?

Reply:
I expect them mainly to be used in linters/analyzers, where the position of certain tokens, and whether they are split to multiple lines or not, grouped, etc... is useful. This is much more important if they want to provide code actions/suggestions, to calculate precisely where to insert new code.

I also think it aligns with our philosophy of "providing a toolbox" for the user and give them the freedom to combine/configure it. Additionally, it doesn't hurt on our side to provide it. It is (almost) zero cost if not used.

Reply:
Normally those use cases would require the CST - which is often the main reason to provide it. The AST wouldn't include them because it is *abstract*. However I agree that if we can provide, we should, on the basis of providing the most powerful thing we can.

Reply:
Additionally, IIRC, we talked about using the nanopass mechanism to further select what is significant vs which is not, to carry over to L1, as it is not as straightforward as just selecting punctuation/keywords, since some of them are significant (`abstract`) and some are not (`pragma` or `function`), so that we are not burdened by it in later passes.

In that sense, this AST abstracts away trivia, and further passes (L1, L2, ...) abstract away other metadata as needed. WDYT?